process/block/preprocess/transactions.go

```diff
@@ -220,65 +232,249 @@ func (txs *transactions) ProcessBlockTransactions(
+		shouldSkipMiniblock := miniBlock.SenderShardID == txs.shardCoordinator.SelfId() || !txs.isMiniBlockCorrect(miniBlock.Type)
+		if shouldSkipMiniblock {
+			continue
+		}
 		if miniBlock.Type != txs.blockType {
+			return nil, fmt.Errorf("%w: block type: %s, sender shard id: %d, receiver shard id: %d",
+				process.ErrInvalidMiniBlockType,
+				miniBlock.Type,
+				miniBlock.SenderShardID,
+				miniBlock.ReceiverShardID)
+		}
+
+		txsFromMiniBlock, err := txs.computeTxsFromMiniBlock(miniBlock)
+		if err != nil {
+			return nil, err
+		}
+
+		allTxs = append(allTxs, txsFromMiniBlock...)
+	}
+
+	return allTxs, nil
+}
+
+func (txs *transactions) computeTxsFromMe(body block.Body) ([]*txcache.WrappedTransaction, error) {
+
+	allTxs := make([]*txcache.WrappedTransaction, 0, process.MaxItemsInBlock)
+
+	for _, miniBlock := range block.MiniBlockSlice(body) {
+		shouldSkipMiniblock := miniBlock.SenderShardID != txs.shardCoordinator.SelfId() || !txs.isMiniBlockCorrect(miniBlock.Type)
+		if shouldSkipMiniblock {
 			continue
 		}
 
-		gasConsumedByMiniBlockInSenderShard := uint64(0)
-		gasConsumedByMiniBlockInReceiverShard := uint64(0)
+		txsFromMiniBlock, err := txs.computeTxsFromMiniBlock(miniBlock)
+		if err != nil {
+			return nil, err
+		}
 
-		for j := 0; j < len(miniBlock.TxHashes); j++ {
-			if !haveTime() {
-				return process.ErrTimeIsOut
-			}
+		allTxs = append(allTxs, txsFromMiniBlock...)
+	}
+
+	return allTxs, nil
+}
+
+func (txs *transactions) computeTxsFromMiniBlock(miniBlock *block.MiniBlock) ([]*txcache.WrappedTransaction, error) {
+	txsFromMiniBlock := make([]*txcache.WrappedTransaction, 0, process.MaxItemsInBlock)
+
+	for i := 0; i < len(miniBlock.TxHashes); i++ {
+		txHash := miniBlock.TxHashes[i]
+		txs.txsForCurrBlock.mutTxsForBlock.RLock()
+		txInfoFromMap := txs.txsForCurrBlock.txHashAndInfo[string(txHash)]
+		txs.txsForCurrBlock.mutTxsForBlock.RUnlock()
+
+		if txInfoFromMap == nil || txInfoFromMap.tx == nil {
+			log.Debug("missing transaction in computeTxsFromMiniBlock", "type", miniBlock.Type, "txHash", txHash)
+			return nil, process.ErrMissingTransaction
+		}
+
+		tx, ok := txInfoFromMap.tx.(*transaction.Transaction)
+		if !ok {
+			return nil, process.ErrWrongTypeAssertion
+		}
+
+		calculatedSenderShardId, err := txs.getShardFromAddress(tx.GetSndAddress())
+		if err != nil {
+			return nil, err
+		}
+
+		calculatedReceiverShardId, err := txs.getShardFromAddress(tx.GetRecvAddress())
+		if err != nil {
+			return nil, err
+		}
+
+		wrappedTx := &txcache.WrappedTransaction{
+			Tx:              tx,
+			TxHash:          txHash,
+			SenderShardID:   calculatedSenderShardId,
+			ReceiverShardID: calculatedReceiverShardId,
+		}
+
+		txsFromMiniBlock = append(txsFromMiniBlock, wrappedTx)
+	}
+
+	return txsFromMiniBlock, nil
+}
+
+func (txs *transactions) getShardFromAddress(address []byte) (uint32, error) {
+	isEmptyAddress := bytes.Equal(address, make([]byte, txs.addressConverter.AddressLen()))
```

Comment:
we should make the empty address as a global const, to not reallocate every time.

Reply:
Done